crates/trie/sparse/src/state.rs

```diff
@@ -20,6 +20,40 @@ use reth_trie_common::{
 };
 use tracing::trace;
 
+/// Provides type-safe re-use of cleared [`SparseStateTrie`]s, which helps to save allocations
+/// across payload runs.
+#[derive(Debug)]
+pub struct ClearedSparseStateTrie<
+    A = SerialSparseTrie, // Account trie implementation
+    S = SerialSparseTrie, // Storage trie implementation
+>(SparseStateTrie<A, S>);
+
+impl<A, S> From<SparseStateTrie<A, S>> for ClearedSparseStateTrie<A, S>
+where
+    A: SparseTrieInterface,
+    S: SparseTrieInterface,
+{
+    fn from(mut trie: SparseStateTrie<A, S>) -> Self {
+        trie.state = trie.state.clear();
+        trie.cleared_storages.extend(trie.storages.drain().map(|(_, trie)| trie.clear()));
+        trie.revealed_account_paths.clear();
+        trie.revealed_storage_paths.clear();
+        trie.account_rlp_buf.clear();
+        Self(trie)
+    }
+}
+
+impl<A, S> ClearedSparseStateTrie<A, S>
+where
+    A: SparseTrieInterface + Default,
+    S: SparseTrieInterface + Default,
+{
+    /// Returns the cleared `SparseStateTrie`, consuming this instance.
+    pub fn unwrap(self) -> SparseStateTrie<A, S> {
```

Comment:
I'd like to name this into_inner, so that this cant be mistaken for a fallible Option::unwrap